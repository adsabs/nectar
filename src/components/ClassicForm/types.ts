import { SolrSort } from '@/api/models';

export type LogicChoice = 'and' | 'or' | 'boolean';
export type CollectionChoice = 'astronomy' | 'physics' | 'general' | 'earthscience';
export type PropertyChoice = 'refereed-only' | 'articles-only';

export interface IClassicFormState {
  limit: CollectionChoice[];
  author: string;
  logic_author: LogicChoice;
  object: string;
  logic_object: LogicChoice;
  pubdate_start: string;
  pubdate_end: string;
  title: string;
  logic_title: LogicChoice;
  abstract_keywords: string;
  logic_abstract_keywords: LogicChoice;
  property: PropertyChoice[];
  bibstems: string;
  sort: SolrSort[];
}

export interface IRawClassicFormState {
  limit: string[];
  author: string;
  logic_author: string;
  object: string;
  logic_object: string;
  pubdate_start: string;
  pubdate_end: string;
  title: string;
  logic_title: string;
  abstract_keywords: string;
  logic_abstract_keywords: string;
  property: string[];
  bibstems: string;
  sort: string[];
}
