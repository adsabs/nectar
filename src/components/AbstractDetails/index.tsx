export * from './AbstractDetails';
export * from './AbstractMetadata';
