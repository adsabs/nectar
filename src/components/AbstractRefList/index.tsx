export * from './AbstractRefList';
export * from './AbstractRefLayout';
