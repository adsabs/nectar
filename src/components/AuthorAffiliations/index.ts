export * from './AuthorAffiliations';
export * from './ErrorMessage';
