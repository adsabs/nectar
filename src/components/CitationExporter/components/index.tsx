export * from './AuthorCutoffField';
export * from './DescriptionCollapse';
export * from './FormatSelect';
export * from './JournalFormatSelect';
export * from './KeyFormatInput';
export * from './MaxAuthorsField';
