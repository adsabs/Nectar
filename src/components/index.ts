export * from './Dropdown';
export * from './Footer';
export * from './Layout';
export * from './ListBox';
export * from './NavBar';
export * from './NumFound';
export * from './ResultList';
export * from './SearchBar';
export * from './SearchExamples';
export * from './Sort';
