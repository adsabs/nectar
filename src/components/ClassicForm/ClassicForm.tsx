import {
  Alert,
  AlertDescription,
  AlertIcon,
  AlertTitle,
  Button,
  Checkbox,
  CheckboxGroup,
  Code,
  Flex,
  FormControl,
  FormHelperText,
  FormLabel,
  HStack,
  Input,
  InputGroup,
  InputLeftElement,
  Radio,
  RadioGroup,
  Stack,
  Text,
  Textarea,
  VisuallyHidden,
} from '@chakra-ui/react';
import { CalendarIcon } from '@chakra-ui/icons';

import { APP_DEFAULTS } from '@/config';
import { useErrorMessage } from '@/lib/useErrorMessage';
import { useIsClient } from '@/lib/useIsClient';
import { useRouter } from 'next/router';
import PT from 'prop-types';
import { FormEventHandler, useMemo } from 'react';
import { Control, Controller, useForm, UseFormRegisterReturn, useWatch } from 'react-hook-form';
import { getSearchQuery } from './helpers';
import { IClassicFormState, IRawClassicFormState } from './types';

import { solrSortOptions } from '@/components/Sort/model';
import { SimpleLink } from '@/components/SimpleLink';
import { BibstemPicker } from '@/components/BibstemPicker';
import { Sort } from '@/components/Sort';
import { Expandable } from '@/components/Expandable';
import { SimpleCopyButton } from '@/components/CopyButton';
import { normalizeSolrSort } from '@/utils/common/search';
import { SolrSort, SolrSortField } from '@/api/models';

const propTypes = {
  ssrError: PT.string,
};

export const defaultClassicFormState: IClassicFormState = {
  limit: ['astronomy'],
  author: '',
  logic_author: 'and',
  object: '',
  logic_object: 'and',
  pubdate_start: '',
  pubdate_end: '',
  title: '',
  logic_title: 'and',
  abstract_keywords: '',
  logic_abstract_keywords: 'and',
  property: [],
  bibstems: '',
  sort: APP_DEFAULTS.SORT,
};

export interface IClassicFormProps {
  ssrError: string;
}

export const ClassicForm = (props: IClassicFormProps) => {
  const isClient = useIsClient();
  const router = useRouter();
  const [queryError, setQueryError] = useErrorMessage<string>(props.ssrError);

  const { register, control, handleSubmit } = useForm<IClassicFormState>({
    defaultValues: defaultClassicFormState,
  });

  const formSubmit: FormEventHandler = (e) => {
    e.preventDefault();

    void handleSubmit((params) => {
      try {
        void router.push({ pathname: '/search', search: getSearchQuery(params) });
      } catch (e) {
        setQueryError((e as Error)?.message);
      }
    })(e);
  };

  return (
    <form method="post" action={router.route} onSubmit={formSubmit} aria-describedby="form-title">
      <Stack direction="column" spacing={5} aria-describedby="form-title">
        <VisuallyHidden as="h2" id="form-title">
          Classic Form
        </VisuallyHidden>

        {/* Collection selection */}
        <FormControl as="fieldset">
          <FormLabel as="legend">Collection</FormLabel>
          <CheckboxGroup defaultValue={['astronomy']}>
            <HStack spacing="6">
              <Checkbox value="astronomy" {...register('limit')}>
                Astronomy
              </Checkbox>
              <Checkbox value="physics" {...register('limit')}>
                Physics
              </Checkbox>
              <Checkbox value="general" {...register('limit')}>
                General
              </Checkbox>
              <Checkbox value="earthscience" {...register('limit')}>
                Earth Science
              </Checkbox>
            </HStack>
          </CheckboxGroup>
        </FormControl>

        {/* Author text area */}
        <FormControl>
          <Flex direction="row" justifyContent="space-between">
            <FormLabel htmlFor={'author'}>Author</FormLabel>
            <LogicRadios variant="andor" radioProps={register('logic_author')} />
          </Flex>
          <Textarea
            {...register('author')}
            as="textarea"
            id="author"
            rows={3}
            placeholder={`Smith, John A\nSmith, Jane B`}
          />
          <FormHelperText lineHeight="5">
            <Text>Author names, enter (Last, First M) one per line.</Text>
            <Text fontWeight="bold">Example Operators:</Text>
            <Text>
              Use <Code>-</Code> to filter out an author. (Ex: <Code>-Smith, John</Code>)
            </Text>
            <Text>
              Use <Code>=</Code> to restrict name expansion. For example <Code>=Smith, Jim</Code> will match &#34;Smith,
              Jim&#34; but not &#34;Smith, James&#34;.
            </Text>
            <Text>
              Surround name with <Code>^ $</Code> to match papers with only one particular author. (Ex:{' '}
              <Code>^Smith, J$</Code>)
            </Text>
            <Text>
              <SimpleLink href="https://ui.adsabs.harvard.edu/help/search/search-syntax#author">Learn More</SimpleLink>
            </Text>
          </FormHelperText>
        </FormControl>

        {/* Object text area */}
        <FormControl>
          <Flex direction="row" justifyContent="space-between">
            <FormLabel htmlFor="object">Object</FormLabel>
            <LogicRadios variant="andor" radioProps={register('logic_object')} />
          </Flex>
          <Textarea
            {...register('object')}
            as="textarea"
            id="object"
            rows={3}
            placeholder={`M 31\nHD 187642\nSgr A*`}
          />
          <FormHelperText>SIMBAD object search, one per line.</FormHelperText>
        </FormControl>

        {/* Publication dates */}
        <Stack
          direction={['column', 'row']}
          justifyContent="space-evenly"
          spacing={5}
          role="group"
          aria-labelledby="publication-group"
          data-testid="publication-dates"
        >
          <FormControl>
            <VisuallyHidden id="publication-group">Publication</VisuallyHidden>
            <FormLabel>Publication Date Start</FormLabel>
            <InputGroup>
              <InputLeftElement pointerEvents="none">
                <CalendarIcon color="gray.300" />
              </InputLeftElement>
              <Input placeholder="YYYY/MM" type="text" {...register('pubdate_start')} />
            </InputGroup>
            <FormHelperText>Ex: &#34;2011/04&#34;</FormHelperText>
          </FormControl>
          <FormControl>
            <FormLabel>Publication Date End</FormLabel>
            <InputGroup>
              <InputLeftElement pointerEvents="none">
                <CalendarIcon color="gray.300" />
              </InputLeftElement>
              <Input placeholder="YYYY/MM" {...register('pubdate_end')} />
            </InputGroup>
            <FormHelperText>Ex: &#34;2014/12&#34;</FormHelperText>
          </FormControl>
        </Stack>

        {/* Title */}
        <FormControl>
          <Flex direction="row" justifyContent="space-between">
            <FormLabel>Title</FormLabel>
            <LogicRadios variant="all" radioProps={register('logic_title')} />
          </Flex>
          <Input {...register('title')} />
          <FormHelperText>Ex: &#34;Content of the Future in the ADS&#34;</FormHelperText>
        </FormControl>

        {/* Abstract / Keywords */}
        <FormControl>
          <Flex direction="row" justifyContent="space-between">
            <FormLabel>Abstract / Keywords</FormLabel>
            <LogicRadios variant="all" radioProps={register('logic_abstract_keywords')} />
          </Flex>
          <Input {...register('abstract_keywords')} />
          <FormHelperText>Ex: &#34;Dark Energy&#34;</FormHelperText>
        </FormControl>

        {/* Property */}
        <FormControl as="fieldset">
          <FormLabel as="legend">Property</FormLabel>
          <CheckboxGroup>
            <HStack spacing={4}>
              <Checkbox fontWeight="bold" value="refereed-only" {...register('property')}>
                Refereed only
              </Checkbox>
              <Checkbox fontWeight="bold" value="articles-only" {...register('property')}>
                Articles only
              </Checkbox>
            </HStack>
          </CheckboxGroup>
        </FormControl>

        {/* BibstemPicker */}
        {isClient ? (
          <Controller
            name="bibstems"
            control={control}
            render={({ field }) => (
              <FormControl>
                <VisuallyHidden id="bibstem">Publications</VisuallyHidden>
                <FormLabel htmlFor="bibstem-picker">Publications</FormLabel>
                <BibstemPicker isMultiple onChange={(items) => field.onChange(items)} id="bibstem-picker" />
                <FormHelperText>Ex. &#34;A&amp;A&#34; or &#34;-A&amp;A&#34;</FormHelperText>
              </FormControl>
            )}
          />
        ) : (
          <FormControl>
            <FormLabel>Publication</FormLabel>
            <Input placeholder="Publication" {...register('bibstems')} />
            <FormHelperText>Start typing to search journal database (ex. &#34;ApJ&#34;)</FormHelperText>
          </FormControl>
        )}

        {/* Sort */}
        <FormControl>
          <FormLabel>Sort</FormLabel>
          <Controller
            name="sort"
            control={control}
            render={({ field }) => (
              <Sort<SolrSort, SolrSortField>
                name={field.name}
                onChange={(sort) => field.onChange(normalizeSolrSort(sort))}
                sort={field.value[0]}
                options={solrSortOptions}
                hiddenInput={{ name: 'sort', value: field.value.join(',') }}
                innerSelectProps={{
                  onBlur: field.onBlur,
                }}
                fullWidth
                useNativeWhenNoJs
              />
            )}
          />
        </FormControl>

        {/* Error processing form */}
        {queryError && (
          <Alert status="error">
            <AlertIcon />
            <AlertTitle>There was a problem parsing input!</AlertTitle>
            <AlertDescription>{queryError}</AlertDescription>
          </Alert>
        )}

        <FormControl>
          <Button type="submit">Search</Button>
        </FormControl>

        <CurrentQuery control={control} />
      </Stack>
    </form>
  );
};

ClassicForm.propTypes = propTypes;

const LogicRadios = (props: { variant: 'andor' | 'all'; radioProps: UseFormRegisterReturn }) => {
  const { variant, radioProps } = props;
  return (
    <RadioGroup defaultValue="and">
      <Stack spacing={4} direction="row">
        <Radio size="sm" {...radioProps} value="and">
          And
        </Radio>
        <Radio size="sm" {...radioProps} value="or">
          Or
        </Radio>
        {variant === 'all' && (
          <Radio size="sm" {...radioProps} value="boolean">
            Boolean
          </Radio>
        )}
      </Stack>
    </RadioGroup>
  );
};

const CurrentQuery = (props: { control: Control<IClassicFormState> }) => {
  const { control } = props;
  const values = useWatch<IClassicFormState>({ control });
  const query = useMemo(() => {
    try {
      return new URLSearchParams(getSearchQuery(values as IRawClassicFormState)).get('q');
    } catch (e) {
      return <Text color="red.500">{(e as Error)?.message}</Text>;
    }
  }, [values]);

  return (
    <Expandable
      title="Generated Query"
      description={
        <HStack>
          <Code>{query}</Code>
          {typeof query === 'string' ? <SimpleCopyButton text={query} /> : null}
        </HStack>
      }
    ></Expandable>
  );
};
