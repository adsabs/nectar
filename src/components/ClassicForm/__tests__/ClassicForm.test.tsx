import { createServerListenerMocks, render } from '@/test-utils';
import { ClassicForm } from '@/components/ClassicForm';
import { expect, test, TestContext, vi } from 'vitest';
import { rest } from 'msw';
import { queryByAttribute } from '@testing-library/dom';

const router = {
  pathname: '/',
  push: vi.fn(),
  asPath: '/',
};

vi.mock('next/router', () => ({
  useRouter: () => router,
}));

test('renders without crashing', () => render(<ClassicForm />));

test('has correct initial values', () => {
  const { getByTestId, getByRole, getAllByRole } = render(<ClassicForm />);

  // limit query checkboxes
  expect(getByRole('checkbox', { name: /Astronomy/i })).toBeChecked();
  expect(getByRole('checkbox', { name: /Physics/i })).not.toBeChecked();
  expect(getByRole('checkbox', { name: /General/i })).not.toBeChecked();
  expect(getByRole('checkbox', { name: /Earth Science/i })).not.toBeChecked();

  // property checkboxes
  expect(getByRole('checkbox', { name: /Refereed only/i })).not.toBeChecked();
  expect(getByRole('checkbox', { name: /Articles only/i })).not.toBeChecked();

  // radio buttons
  getAllByRole('radio', { name: /And/i }).forEach((radio) => expect(radio).toBeChecked());
  getAllByRole('radio', { name: /Or/i }).forEach((radio) => expect(radio).not.toBeChecked());
  getAllByRole('radio', { name: /Boolean/i }).forEach((radio) => expect(radio).not.toBeChecked());

  // textboxes should all be empty
  expect(getAllByRole('textbox')).toHaveLength(6);
  getAllByRole('textbox').forEach((textbox) => expect(textbox).toHaveValue(''));

  expect(getByTestId('bibstem-picker-hidden-input')).toHaveValue('');
  expect(getByTestId('sort-select')).toHaveValue('score');

  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('checkboxes updates query properly', async () => {
  const { getByTestId, getByRole, user } = render(<ClassicForm />);

  // limit query checkboxes
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getByRole('checkbox', { name: /Astronomy/i }));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getByRole('checkbox', { name: /Physics/i }));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getByRole('checkbox', { name: /General/i }));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getByRole('checkbox', { name: /Earth Science/i }));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();

  // property checkboxes
  await user.click(getByRole('checkbox', { name: /Refereed only/i }));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getByRole('checkbox', { name: /Articles only/i }));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('author textbox updates query', async () => {
  const { getByTestId, getByRole, user } = render(<ClassicForm />);
  const getRadio = (name: string, logic: 'And' | 'Or' | 'Boolean') =>
    getByRole('radio', {
      name: (aName, el) => el.getAttribute('name') === `logic_${name}` && aName === logic,
    });

  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Author/i }), 'Smith, J');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Author/i }), '\nSmith, A\nJames, B');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Author/i }), '\n-Smith, John\n=Smith, John\n^Smith, J$');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getRadio('author', 'Or'));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('object textbox updates query', async () => {
  const { getByTestId, getByRole, user } = render(<ClassicForm />);
  const getRadio = (name: string, logic: 'And' | 'Or' | 'Boolean') =>
    getByRole('radio', {
      name: (aName, el) => el.getAttribute('name') === `logic_${name}` && aName === logic,
    });

  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Object/i }), 'foo');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Object/i }), '\nbar\nbaz');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getRadio('object', 'Or'));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('title textbox updates query', async () => {
  const { getByTestId, getByRole, user } = render(<ClassicForm />);
  const getRadio = (name: string, logic: 'And' | 'Or' | 'Boolean') =>
    getByRole('radio', {
      name: (aName, el) => el.getAttribute('name') === `logic_${name}` && aName === logic,
    });

  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Title/i }), 'foo');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Title/i }), ' bar baz');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getRadio('title', 'Or'));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getRadio('title', 'Boolean'));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('abstract textbox updates query', async () => {
  const { getByTestId, getByRole, user } = render(<ClassicForm />);
  const getRadio = (name: string, logic: 'And' | 'Or' | 'Boolean') =>
    getByRole('radio', {
      name: (aName, el) => el.getAttribute('name') === `logic_${name}` && aName === logic,
    });

  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Abstract/i }), 'foo');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.type(getByRole('textbox', { name: /Abstract/i }), ' bar baz');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getRadio('abstract_keywords', 'Or'));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
  await user.click(getRadio('abstract_keywords', 'Boolean'));
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('publications textbox updates query', async ({ server }: TestContext) => {
  const { onRequest } = createServerListenerMocks(server);

  server.use(
    rest.get<{ term: string }>('*/api/bibstems/:term', (req, res, ctx) => {
      if (req.params.term === 'a') {
        return res(ctx.json([{ value: 'ApJ', label: ['The Astrophysical Journal'] }]));
      } else if (req.params.term === 'b') {
        return res(ctx.json([{ value: 'BGeo', label: ['Biogeosciences'] }]));
      }
      return res(ctx.json([]));
    }),
  );

  const { getByTestId, getAllByTestId, container, user, getByRole } = render(<ClassicForm />);

  // selects from the default list
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '{Enter}');
  expect(getAllByTestId('pill').map((pill) => pill.textContent)).toMatchSnapshot();
  expect(getByTestId('bibstem-picker-hidden-input').getAttribute('value')).toMatchSnapshot();

  // selects from the default list again
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '{Enter}');
  expect(getAllByTestId('pill').map((pill) => pill.textContent)).toMatchSnapshot();
  expect(getByTestId('bibstem-picker-hidden-input').getAttribute('value')).toMatchSnapshot();

  // select one from async list
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), 'a');
  expect(queryByAttribute('id', container, /.*bibstem-picker-listbox/i)).toMatchSnapshot();
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '{Enter}');
  expect(getByTestId('bibstem-picker-hidden-input').getAttribute('value')).toMatchSnapshot();

  // select one from async list but negates it
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '-b');
  expect(queryByAttribute('id', container, /.*bibstem-picker-listbox/i)).toMatchSnapshot();
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '{Enter}');
  expect(getByTestId('bibstem-picker-hidden-input').getAttribute('value')).toMatchSnapshot();

  // creates a custom bibstem
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), 'e');
  expect(queryByAttribute('id', container, /.*bibstem-picker-listbox/i)).toMatchSnapshot();
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '{Enter}');
  expect(getByTestId('bibstem-picker-hidden-input').getAttribute('value')).toMatchSnapshot();

  // creates a negated custom bibstem
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '-e');
  expect(queryByAttribute('id', container, /.*bibstem-picker-listbox/i)).toMatchSnapshot();
  await user.type(getByRole('combobox', { name: /Bibstem Picker/i }), '{Enter}');
  expect(getByTestId('bibstem-picker-hidden-input').getAttribute('value')).toMatchSnapshot();

  // check the server mock to make sure we made the right calls
  expect(onRequest.mock.calls).toHaveLength(4);
  expect(onRequest.mock.calls.map((call) => call[0].url.pathname)).toMatchSnapshot();

  // finally check the generated query to confirm it's been updated correctly
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});

test('pubdate updates the query', async () => {
  const { getByTestId, user, getByRole } = render(<ClassicForm />);

  // set the start date
  await user.type(getByRole('textbox', { name: /Publication Date Start/i }), '2020/01');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();

  // set the end date
  await user.type(getByRole('textbox', { name: /Publication Date End/i }), '2021/01');
  expect(getByTestId('generated-query').textContent).toMatchSnapshot();
});
