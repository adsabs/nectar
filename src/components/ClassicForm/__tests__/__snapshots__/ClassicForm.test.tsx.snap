// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`abstract textbox updates query 1`] = `"collection:(astronomy)"`;

exports[`abstract textbox updates query 2`] = `"collection:(astronomy) abs:(foo)"`;

exports[`abstract textbox updates query 3`] = `"collection:(astronomy) abs:(foo bar baz)"`;

exports[`abstract textbox updates query 4`] = `"collection:(astronomy) abs:(foo OR bar OR baz)"`;

exports[`abstract textbox updates query 5`] = `"collection:(astronomy) abs:(foo bar baz)"`;

exports[`author textbox updates query 1`] = `"collection:(astronomy)"`;

exports[`author textbox updates query 2`] = `"collection:(astronomy) author:("Smith, J")"`;

exports[`author textbox updates query 3`] = `"collection:(astronomy) author:("Smith, J" "Smith, A" "James, B")"`;

exports[`author textbox updates query 4`] = `"collection:(astronomy) author:("Smith, J" OR "Smith, A" OR "James, B" OR -"Smith, John" OR ="Smith, John" OR "^Smith, J") author_count:1"`;

exports[`author textbox updates query 5`] = `"collection:(astronomy) author:("Smith, J" OR "Smith, A" OR "James, B" OR -"Smith, John" OR ="Smith, John" OR "^Smith, J") author_count:1"`;

exports[`checkboxes updates query properly 1`] = `"collection:(astronomy)"`;

exports[`checkboxes updates query properly 2`] = `""`;

exports[`checkboxes updates query properly 3`] = `"collection:(physics)"`;

exports[`checkboxes updates query properly 4`] = `"collection:(physics general)"`;

exports[`checkboxes updates query properly 5`] = `"collection:(physics general earthscience)"`;

exports[`checkboxes updates query properly 6`] = `"collection:(physics general earthscience) property:(refereed)"`;

exports[`checkboxes updates query properly 7`] = `"collection:(physics general earthscience) property:(refereed article)"`;

exports[`has correct initial values 1`] = `"collection:(astronomy)"`;

exports[`object textbox updates query 1`] = `"collection:(astronomy)"`;

exports[`object textbox updates query 2`] = `"collection:(astronomy) object:(foo)"`;

exports[`object textbox updates query 3`] = `"collection:(astronomy) object:(foo bar baz)"`;

exports[`object textbox updates query 4`] = `"collection:(astronomy) object:(foo OR bar OR baz)"`;

exports[`pubdate updates the query 1`] = `"collection:(astronomy) pubdate:[2020-01 TO 9999-12]"`;

exports[`pubdate updates the query 2`] = `"collection:(astronomy) pubdate:[2020-01 TO 2021-01]"`;

exports[`publications textbox updates query 1`] = `
[
  "LRCA",
]
`;

exports[`publications textbox updates query 2`] = `"LRCA"`;

exports[`publications textbox updates query 3`] = `
[
  "LRCA",
  "PhRvL",
]
`;

exports[`publications textbox updates query 4`] = `"LRCA,PhRvL"`;

exports[`publications textbox updates query 5`] = `
<div
  class=" css-h3hx5y-menu"
  id="react-select-bibstem-picker-listbox"
>
  <div
    class=" css-1n6sfyn-MenuList"
  >
    <div
      aria-disabled="false"
      class=" css-1kmzzr7-option"
      id="react-select-bibstem-picker-option-0"
      tabindex="-1"
    >
      <div
        class="chakra-stack css-1oeb4ru"
        data-testid="option"
      >
        <p
          class="chakra-text css-in3yi3"
        >
          ApJ
        </p>
        <p
          class="chakra-text css-13xnkip"
        >
          The Astrophysical Journal
        </p>
      </div>
    </div>
    <div
      aria-disabled="false"
      class=" css-16grddd-option"
      id="react-select-bibstem-picker-option-1"
      tabindex="-1"
    >
      <div
        class="chakra-stack css-1oeb4ru"
        data-testid="option"
      >
        <p
          class="chakra-text css-in3yi3"
        >
          a
        </p>
        <p
          class="chakra-text css-13xnkip"
        >
          Custom Journal? insert "a"
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`publications textbox updates query 6`] = `"LRCA,PhRvL,ApJ"`;

exports[`publications textbox updates query 7`] = `
<div
  class=" css-h3hx5y-menu"
  id="react-select-bibstem-picker-listbox"
>
  <div
    class=" css-1n6sfyn-MenuList"
  >
    <div
      aria-disabled="false"
      class=" css-1kmzzr7-option"
      id="react-select-bibstem-picker-option-0"
      tabindex="-1"
    >
      <div
        class="chakra-stack css-1oeb4ru"
        data-testid="option"
      >
        <p
          class="chakra-text css-in3yi3"
        >
          BGeo
        </p>
        <p
          class="chakra-text css-13xnkip"
        >
          Biogeosciences
        </p>
      </div>
    </div>
    <div
      aria-disabled="false"
      class=" css-16grddd-option"
      id="react-select-bibstem-picker-option-1"
      tabindex="-1"
    >
      <div
        class="chakra-stack css-1oeb4ru"
        data-testid="option"
      >
        <p
          class="chakra-text css-in3yi3"
        >
          -b
        </p>
        <p
          class="chakra-text css-13xnkip"
        >
          Custom Journal? insert "-b"
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`publications textbox updates query 8`] = `"LRCA,PhRvL,ApJ,-BGeo"`;

exports[`publications textbox updates query 9`] = `
<div
  class=" css-h3hx5y-menu"
  id="react-select-bibstem-picker-listbox"
>
  <div
    class=" css-1n6sfyn-MenuList"
  >
    <div
      aria-disabled="false"
      class=" css-1kmzzr7-option"
      id="react-select-bibstem-picker-option-0"
      tabindex="-1"
    >
      <div
        class="chakra-stack css-1oeb4ru"
        data-testid="option"
      >
        <p
          class="chakra-text css-in3yi3"
        >
          e
        </p>
        <p
          class="chakra-text css-13xnkip"
        >
          Custom Journal? insert "e"
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`publications textbox updates query 10`] = `"LRCA,PhRvL,ApJ,-BGeo,e"`;

exports[`publications textbox updates query 11`] = `
<div
  class=" css-h3hx5y-menu"
  id="react-select-bibstem-picker-listbox"
>
  <div
    class=" css-1n6sfyn-MenuList"
  >
    <div
      aria-disabled="false"
      class=" css-1kmzzr7-option"
      id="react-select-bibstem-picker-option-0"
      tabindex="-1"
    >
      <div
        class="chakra-stack css-1oeb4ru"
        data-testid="option"
      >
        <p
          class="chakra-text css-in3yi3"
        >
          -e
        </p>
        <p
          class="chakra-text css-13xnkip"
        >
          Custom Journal? insert "-e"
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`publications textbox updates query 12`] = `"LRCA,PhRvL,ApJ,-BGeo,e,-e"`;

exports[`publications textbox updates query 13`] = `
[
  "/api/bibstems/a",
  "/api/bibstems/b",
  "/api/bibstems/e",
  "/api/bibstems/e",
]
`;

exports[`publications textbox updates query 14`] = `"collection:(astronomy) bibstem:(LRCA OR PhRvL OR ApJ OR e) -bibstem:(BGeo OR e)"`;

exports[`title textbox updates query 1`] = `"collection:(astronomy)"`;

exports[`title textbox updates query 2`] = `"collection:(astronomy) title:(foo)"`;

exports[`title textbox updates query 3`] = `"collection:(astronomy) title:(foo bar baz)"`;

exports[`title textbox updates query 4`] = `"collection:(astronomy) title:(foo OR bar OR baz)"`;

exports[`title textbox updates query 5`] = `"collection:(astronomy) title:(foo bar baz)"`;
